internal/mirror/mirror.go

```diff
@@ -0,0 +1,276 @@
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/google/exposure-notifications-server/internal/database"
+	"github.com/google/exposure-notifications-server/internal/mirror/model"
+	"github.com/google/exposure-notifications-server/internal/storage"
+	"github.com/google/exposure-notifications-server/pkg/logging"
+	"go.opencensus.io/trace"
+)
+
+const mirrorLockPrefix = "mirror-lock"
+
+type FileStatus struct {
+	Order        int
+	MirrorFile   *model.MirrorFile
+	DownloadPath string
+	Filename     string
+	Failed       bool
+	Saved        bool
+}
+
+func (f *FileStatus) needsDelete() bool {
+	return f.DownloadPath == ""
+}
+
+func (f *FileStatus) needsDownload() bool {
+	return f.MirrorFile == nil
+}
+
+func (s *Server) downloadIndex(mirror *model.Mirror) ([]string, error) {
+	client := &http.Client{
+		Timeout: s.config.IndexFileDownloadTimeout,
+	}
+	resp, err := client.Get(mirror.IndexFile)
+	if err != nil {
+		return nil, fmt.Errorf("error downloading index file: %w", err)
+	}
+	defer resp.Body.Close()
+	bytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading index file: %w", err)
+	}
+	zipNames := strings.Split(string(bytes), "\n")
+	currentFiles := make([]string, 0, len(zipNames))
+	for _, zipFile := range zipNames {
+		fullZipFile := fmt.Sprintf("%s%s", mirror.ExportRoot, strings.TrimSpace(zipFile))
+		currentFiles = append(currentFiles, fullZipFile)
+	}
+
+	return currentFiles, nil
+}
+
+func (s *Server) handleMirror(ctx context.Context) http.HandlerFunc {
+	logger := logging.FromContext(ctx).Named("mirror.handleMirror")
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		_, span := trace.StartSpan(ctx, "(*mirror.handleMirror).ServeHTTP")
+		defer span.End()
+
+		ctx, cancelFn := context.WithTimeout(ctx, s.config.MaxRuntime)
+		defer cancelFn()
+
+		// Chop of 30 seconds to save state at the end.
+		runtime := s.config.MaxRuntime
+		if runtime > time.Minute {
+			runtime = runtime - (30 * time.Second)
+		}
+		deadline := time.Now().Add(runtime)
+		logger.Infow("mirrow will run until", "deadline", deadline)
+
+		// get all possible mirror candidates
+		mirrors, err := s.mirrorDB.Mirrors(ctx)
+		if err != nil {
+			logger.Errorw("unable to list mirrors", "error", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		blobstore := s.env.Blobstore()
+
+		for _, mirror := range mirrors {
+			if deadlinePassed(deadline) {
+				logger.Warnw("mirror timed out before processing all configurations")
+				break
+			}
+
+			lockID := fmt.Sprintf("%s-%d", mirrorLockPrefix, mirror.ID)
+			unlock, err := s.db.Lock(ctx, lockID, s.config.MaxRuntime)
+			if err != nil {
+				if errors.Is(err, database.ErrAlreadyLocked) {
+					logger.Infow("mirror already locked, skipping", "mirrorID", mirror.ID)
+					continue
+				}
+				logger.Warn(err)
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			defer func() {
```

Comment:
This defer won't run until _all_ things in the loop have finished and the rest of the function finishes executing. I don't think that's what you want?

Reply:
extracted processing to a function so all defers will run.